### PR TITLE
SD card access and Read Only FAT32 file system support

### PR DESCRIPTION
**Pull request by alees24**
**CI failures are expected**; there is no microSD card( image) in ~~either~~ simulation~~or FPGA testing~~ at present.

This PR does the following:

Commit 1:
- Introduces SD card support on top of the SPI driver
- Adds a _manual_ 'check' program called 'sdraw_check' that insists upon manual insertion of a card that may be corrupted by the test, and then performs a number of randomised block writes and reads to check that these occur successfully.

Commit 2:
- Introduces Read Only support for a single FAT32 partition (as SD cards are typically shipped by manufacturers).
- Adds SD card testing to the 'test_runner' regression test in CI to test (i) SD card presence, (ii) the ability to read the card properties, (iii) the ability to read a sample text file from the root directory of a FAT32 partition on the SD card (iv) that it matches against a reference copy.

Whilst 'sdraw_check' may be run against any modern microSD card of suitable capacity _that does not contain data you wish to keep_, the test program in the second commit requires that the card be prepared with a suitable file 'LOREM.IPS' (with that exact name and capitalisation at present). Detailed instructions may be found in the source file `sw/cheri/tests/sdcard_tests.hh`; in outline, modify that file by setting 'emitText = true' before running it for the first time and it will emit the instructions and the sample text that must be used in setting up the microSD card.

Create a single FAT32 partition on the card, using the 'Master Boot Record' scheme and not 'GPT.' Newly-purchased cards should already be suitable although only 32GiB and 64GiB cards have been tried to date.

Still to come, and probably required before we can merge this (to keep the FPGA and simulation consistent):
~~1. DPI model of a SPI-connected SD card; this allows the above test to pass in simulation too.~~

Still on the 'TODO' list - but perhaps may be left - until a later PR are:
~~1. Improved filename matching within directories, including Long FileName support.~~
~~2. A fresh attempt to get multiple block read/write operations working; these are presently implemented using repeated single block reads.~~

### CODE REVIEW

File: sw/cheri/common/sdcard-utils.hh
Comment (marnovandermaas) [inline sw/cheri/common/sdcard-utils.hh:151]: This is nearly the same as the CID read. Can you avoid the code repetition?

File: sw/cheri/tests/sdcard_tests.hh
Comment (marnovandermaas) [inline sw/cheri/tests/sdcard_tests.hh:50]: Is this the same as the one in the sdcard check? Can we re-use them somehow?

File: sw/cheri/tests/sdcard_tests.hh
Comment (marnovandermaas) [inline sw/cheri/tests/sdcard_tests.hh:188]: Nit: remove the commented out code?

File: sw/cheri/checks/sdraw_check.cc
Comment (marnovandermaas) [inline sw/cheri/checks/sdraw_check.cc:164]: Is this function only used for tests, or can it also be used in a check?

**Review by marnovandermaas — COMMENTED** [review]
Thanks for putting this together. I just left a few comments while reading through the code. We'll need to get CI to pass before merging, but this is good enough to share around to people that want to get started with SD cards.

**Comment by GregAC** [comment]
Thanks for this @alees24.

I'm keen we do not spend significant time on this right now so let's not worry about spending time on an SD card DPI model. As we already have the test code it's reasonable for us to include the test in CI to run against the real FPGA.

We also need to consider how to to best use this in sonata-software. It's important we have a working example in there, so where shall `sdcard-utils.hh` live? We could hopefully refactor things so it can work both in the bare metal environment and CHERIoT RTOS environment. I think it's best left in this repository so we don't introduce a new sonata-system -> sonata-software dependency. We have a sonata-software -> sonata-system link though it's done via nix. In order to provide easy code access for people who aren't using the nix environment (i.e. we want a way to build an SD card example in sonata-software without relying on some nix magic to pull the `sdcard-utils.hh` into the right place) either we bring sonata-system in as a sub-module or we just make a copy of it/vendor it in sonata-software. I think I favour the latter as it's just one file and we don't really want the whole of sonata-system as a big submodule.

We could also just say no sd card util at all in sonata-system to simplify things. That would mean no testing of sd card functionality in CI though. I don't think it's that important as it's just a plain SPI device but would mean if we mess up something do with the pinmux/pin mapping it would get flagged.

**Comment by alees24** [comment]
The DPI model already exists and works; sdcard_tests.hh works in sim and it's even possible to play video files to the simulated LCD; that was done a couple of weeks ago. I guess it wasn't clear that I meant only that it needs a little tidying before inclusion, so it's not in this PR.

File: sw/cheri/checks/sdraw_check.cc
Comment (alees24) [reply sw/cheri/checks/sdraw_check.cc:26]: Copied/mimicked code; it's done in other files. I think we even still have a source file that #includes a single file twice, once inside those comments and once outside!

File: sw/cheri/tests/sdcard_tests.hh
Comment (alees24) [reply sw/cheri/tests/sdcard_tests.hh:50]: There are presently 3 copies, and one has a comment about moving this into console.hh; I'll move it there unless you can think of a better home?

**Comment by GregAC** [comment]
> The DPI model already exists and works; 

Awesome. May as well leverage it then.

re: example in sonata-software I think the thing to do is get this PR merged ignoring that issue. We can then deal with that separately.

File: sw/cheri/tests/sdcard_tests.hh
Comment (alees24) [reply sw/cheri/tests/sdcard_tests.hh]: I've moved it to console.hh as proposed.

**Comment by alees24** [comment]
Right, I think this is ready to go; long filename support and directory traversal/matching could use some more testing but it works well enough for our present needs.

The DPI model is now present, and although the FPGA system presently has problems, 'test_runner' should pass on this PR when they have been resolved. For simulation, however, we have an issue.....it needs as 'sd.img' file that is ca. 1GiB (see the instructions I've added in 'doc/guide/sdcard-setup.md') which won't be present when the simulation is running, presumably....suggestions welcome, please!

File: sw/cheri/tests/sdcard_tests.hh
Comment (marnovandermaas) [reply sw/cheri/tests/sdcard_tests.hh]: Thanks, that sounds good to me.

File: sw/cheri/tests/sdcard_tests.hh
Comment (AlexJones0) [inline sw/cheri/tests/sdcard_tests.hh:131]: Just a small error I noticed looking through, shouldn't cause any issues though. 
```suggestion
  SonataPinmux::Sink appspi_clk   = pin_output->get(SonataPinmux::PinSink::appspi_clk);
```

**Comment by alees24** [comment]
I see a few possible options for resolving the simulator situation:
- put the file 'sd.img' in the repository; some people don't like binary files in repositories, and it may need to be quite large; I don't presently know the minimum capacity volume that the code will accept.
- the SD DPI model could be modified to construct a suitable image if there is not one available, or this could be done as a step in the CI build flow; a short sequence of shell commands ought to suffice, and the reference text would need to be extracted from the header file and written as a file on the mounted virtual volume.
- sdcard_tests.hh operates differently on sim c.f. FPGA and does not access a disk image; I'd prefer to avoid this disparity if possible.

Opinions/thoughts, please?

**Comment by marnovandermaas** [comment]
> I see a few possible options for resolving the simulator situation:
> 
>     * put the file 'sd.img' in the repository; some people don't like binary files in repositories, and it may need to be quite large; I don't presently know the minimum capacity volume that the code will accept.
If we need to go done this route, we can use git large file storage.
> 
>     * the SD DPI model could be modified to construct a suitable image if there is not one available, or this could be done as a step in the CI build flow; a short sequence of shell commands ought to suffice, and the reference text would need to be extracted from the header file and written as a file on the mounted virtual volume.
My initial though is that this is best. Would you be able to have a go at this?
> 
>     * sdcard_tests.hh operates differently on sim c.f. FPGA and does not access a disk image; I'd prefer to avoid this disparity if possible.
> 
> 
> Opinions/thoughts, please?

See my opinions inline.

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:93]: In my experience working with SD cards, it's better to actually do the initialization properly because you will find cards in the while that enforce the spec to the letter.

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:107]: Per the spec, a single command should suffice here, do you really need to loop?

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:111]: Supporting a v1 card is not actually hard: if the card does not answer this command then it's v1. Maybe at least print an error message and bail out?

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:95]: I believe that in theory, you are not supposed to go above 400kHz when initializating the card.

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:135]: I believe that once you get a successful answer from CMD58 you can switch to the default clock rate which is 20MHz for SPI mode.

File: sw/cheri/common/sdcard-utils.hh
Comment (pamaury) [inline sw/cheri/common/sdcard-utils.hh:262]: This is actually the opposite: you need 8 clock ticks *after commands*. Per the spec:

> After the last SD Memory Card bus transaction, the host is required, to provide 8 (eight) clock
> cycles for the card to complete the operation before shutting down the clock. Following is a list of the
> various bus transactions:
> •A command with no response. 8 clocks after the host command end bit.
> •A command with response. 8 clocks after the card response end bit.
> •A read data transaction. 8 clocks after the end bit of the last data block.
> •A write data transaction. 8 clocks after the CRC status token.

**Review by marnovandermaas — APPROVED** [review]
Approving this, but let's wait for CI to pass before merging.

File: sw/cheri/common/sdcard-utils.hh
Comment (alees24) [reply sw/cheri/common/sdcard-utils.hh:93]: Thanks for all the feedback; the code was originally a quick prototype just to check that we could use the SD card at all before 1.0, and it was based upon rather ad hoc tutorials/code found online, without having a proper specification.

It's only example code, but perhaps - at our leisure - we should shore it up properly; @GregAC plans to improve the robustness and error detection of the code.

**Comment by alees24** [comment]
> Re the SD card image file, I'd rather not have everyone who wants to check out the sonata-system repository need to fetch a massive file. Is there any reason we need such a large file? Alternatively we could store it compressed (I'm assuming as an empty file system image it'll compress very very well)
> 
> Is there a way to mark an LFS file as 'optional' or something? I.e. a normal clone won't download it and you need another command to actually fetch it.

I remarked earlier that I don't know how small the image can be and still be usable. 1GiB is merely the size I've been using until now, but it may even work fine with as little as 8KiB, if software can create a suitable image. For now it doesn't matter, we've just restricted the testing in simulation; to be addressed later. It's still exercising the SPI and microSD card interface and doing data reads from CID and CSD in sim, just not from simulated flash storage.

**Pull request merged**
